cpp/src/arrow/util/future.cc

```diff
@@ -272,6 +272,8 @@ class ConcreteFutureImpl : public FutureImpl {
         return true;
       case ShouldSchedule::IfUnfinished:
         return !in_add_callback;
+      case ShouldSchedule::IfDifferentExecutor:
+        return !callback_record.options.executor->OwnsThisThread();
```

Comment:
Is this speculative?  Or is it actually needed?

Reply:
I use it in SourceNode